HTML API: Use full parser for html5lib tests

sirreal opened the pull request:
Trac ticket: Core-61646

Previously, the html5lib tests have only run in the fragment parser mode, but in #6977 a full parser was introduced to the HTML API. In this patch, the html5lib tests are running in the full parser mode where applicable.

#7146 is merged here and should be landed first.

```diff
- Tests: 609, Assertions: 405, Skipped: 204.
+ Tests: 1498, Assertions: 787, Skipped: 711.
```

This vastly increases the tests that are used from HTML5lib-tests from 609 to 1498! It's helped to discover two issues:

- https://core.trac.wordpress.org/ticket/61831
- https://core.trac.wordpress.org/ticket/61810

sirreal committed:
HTML5Lib: enable head tests

sirreal committed:
HTML5Lib: Use full parser when context not provided

sirreal committed:
HTML5Lib: Strip doctypes from expected output

Doctypes are not exposed via next_token so cannot be put into the tree

sirreal committed:
HTML5Lib: Ignore tests with known issues

sirreal committed:
HTML5Lib: Handle PI lookalike comments

sirreal committed:
HTML5Lib: Handle funky comments in tree construction

sirreal committed:
PICKME: Bugfix on ?-initial invalid comment texts

sirreal committed:
HTML5Lib: Add special handling for missing html, head, body tags

HTML, HEAD, and BODY tags should always be generated.
This breaks many tests. Add the missing tags to the processed tree so tests pass

sirreal committed:
Finish skipping tests

sirreal committed:
Revert "PICKME: Bugfix on ?-initial invalid comment texts"

This reverts commit e1a4234e8d75cdce3904e3c9d0d745f50111c017.

sirreal committed:
Revert "fixup! PICKME: Bugfix on ?-initial invalid comment texts"

This reverts commit 770acec942a14a6f882a25c30ed030f2b527cf92.

sirreal committed:
Disable some unimplemented tests

sirreal committed:
Read the script-on flag and ignore tests

sirreal committed:
Fix up ignores

sirreal committed:
Test ignores cleanup

sirreal committed:
Lints

sirreal commented on tests/phpunit/tests/html-api/wpHtmlProcessorHtml5lib.php:
```diff
@@ -68,14 +74,43 @@ class Tests_HtmlApi_Html5lib extends WP_UnitTestCase {
+		$auto_generated_html_head_body = "<html>\n  <head>\n  <body>\n\n";
+		$auto_generated_head_body      = "  <head>\n  <body>\n\n";
+		$auto_generated_body           = "  <body>\n\n";
+		if ( str_ends_with( $expected_tree, $auto_generated_html_head_body ) && ! str_ends_with( $processed_tree, $auto_generated_html_head_body ) ) {
+			if ( str_ends_with( $processed_tree, "<html>\n  <head>\n\n" ) ) {
+				$processed_tree = substr_replace( $processed_tree, "  <body>\n\n", -1 );
+			} elseif ( str_ends_with( $processed_tree, "<html>\n\n" ) ) {
+				$processed_tree = substr_replace( $processed_tree, "  <head>\n  <body>\n\n", -1 );
+			} else {
+				$processed_tree = substr_replace( $processed_tree, $auto_generated_html_head_body, -1 );
+			}
+		} elseif ( str_ends_with( $expected_tree, $auto_generated_head_body ) && ! str_ends_with( $processed_tree, $auto_generated_head_body ) ) {
+			if ( str_ends_with( $processed_tree, "<head>\n\n" ) ) {
+				$processed_tree = substr_replace( $processed_tree, "  <body>\n\n", -1 );
+			} else {
+				$processed_tree = substr_replace( $processed_tree, $auto_generated_head_body, -1 );
+			}
+		} elseif ( str_ends_with( $expected_tree, $auto_generated_body ) && ! str_ends_with( $processed_tree, $auto_generated_body ) ) {
+			$processed_tree = substr_replace( $processed_tree, $auto_generated_body, -1 );
+		}
```
`str_ends_with` will need to be replaced with something PHP-7.2 compatible. It's a PHP8 feature.

dmsnell replied:
we polyfill it in WordPress - it should be find here, unless we don't load Wordpress

sirreal committed:
PICKME: Fix infinite loop in skip_script_data

dmsnell committed:
HTML API: Allow any fragment context.

Previously, the fragment parser in WP_HTML_Processor has only allowed
creating a fragment with the `<body>` context. In this patch, any
context node is allowed.

dmsnell commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -3532,7 +3545,13 @@ private function matches(): bool {
+				strlen( $this->sought_tag_name ) !== $this->tag_name_length ||
+				0 !== substr_compare( $this->html, $this->sought_tag_name, $this->tag_name_starts_at, $this->tag_name_length, true )
+			)
+		) {
```
we need this. `next_tag( 'PLAINTEXT' )` stopped on `P`

dmsnell commented:
@sirreal I've added active format reconstruction to the `PLAINTEXT` element, even though it's missing from the spec, but all the major parsers clearly perform this. I've submitted a PR to the HTML spec to add the rule.

https://github.com/whatwg/html/pull/10540

dmsnell commented:
Leaving a note for posterity's sake: adding active format reconstruction to PLAINTEXT is wrong. Adding PLAINTEXT is more complicated than it originally seemed, because even though every remaining token is a character token, if there are any character tokens left, they may trigger active format reconstruction. This is something I'll try and fix

dmsnell committed:
Merge branch 'trunk' into html-api/support-initial-mode

dmsnell commented:
Fragment support extracted into https://github.com/WordPress/wordpress-develop/pull/7141 where I thought I had already done so, sorry!

sirreal commented:
Thanks, I'll revert the changes to support arbitrary fragments here and continue to only support `body`. That work is more complicated and can continue in #7141.

sirreal committed:
Revert "HTML API: Allow any fragment context."

This reverts commit a38568553d8ed2a6046ebe5a583d93cd53d09548.

sirreal committed:
Remove SKIP_HEAD_TESTS code

This is no longer required. The full or fragment parser will be
used as necessary.

sirreal committed:
Remove skip for bug

The "</#" HTML test demonstrates a bug that must be fixed.

sirreal committed:
Add test case for bug

sirreal committed:
Set state to incomplete on short HTML funky comments

A state change was missing when the input is too short to find a comment
closer in an opened funky comment. This fixes a issue where `</#` does
not correctly report incomplete input, but `</# ` does.

sirreal committed:
Merge branch 'html-api/fix-unclosed-comment-bug' into html-api/support-initial-mode

sirreal committed:
Remove unnecessary test skips

sirreal commented on tests/phpunit/tests/html-api/wpHtmlProcessorHtml5lib.php:
```diff
@@ -68,14 +55,43 @@ class Tests_HtmlApi_Html5lib extends WP_UnitTestCase {
-		$this->assertSame( $expected_tree, $processed_tree, "HTML was not processed correctly:\n{$html}" );
+		/*
+		 * The HTML processor does not produce html, head, body tags if the processor does not reach them.
+		 * These should all be produced when reaching the end-of-file.
+		 * For now, append the missing tags when necessary.
+		 *
+		 * @todo remove this section when when the processor handles this.
```
```suggestion
		 * HTML tree construction will always produce these tags, the HTML API does not at this time.
```

sirreal committed:
Improve comment where html, head, body tags are appended.

dmsnell committed:
Merge branch 'trunk' into html-api/support-initial-mode

dmsnell reviewed (approved):
This is a nice improvement in how many tests run, and it's not trivial to fix the test runner. Thanks again @sirreal!

This is a tests-only change.

dmsnell commented:
Merged in [[58859]](https://core.trac.wordpress.org/changeset/58859)
https://github.com/wordpress/wordpress-develop/commit/9928cd6bcb35d14a700e3930355b6703f0cb4dc8
